src/main/kotlin/uk/gov/justice/digital/hmpps/authorizationserver/service/ClientsService.kt

```diff
@@ -91,6 +97,29 @@ class ClientsService(
   fun findClientByClientId(clientId: String) =
     clientRepository.findClientByClientId(clientId)
       ?: throw ClientNotFoundException(Client::class.simpleName, clientId)
+
+  @Transactional
+  fun duplicate(clientId: String): ClientCredentialsRegistrationResponse {
+    val clientsByBaseClientId = clientIdService.findByBaseClientId(clientId)
+    if (clientsByBaseClientId.isEmpty()) {
+      throw ClientNotFoundException(Client::class.simpleName, clientId)
+    }
+
+    val client = clientsByBaseClientId.last()
+    val registeredClient = registeredClientRepository.findByClientId(client.clientId)
+    val registeredClientBuilder = RegisteredClient.from(registeredClient)
+
+    val externalClientSecret = oAuthClientSecret.generate()
+    val duplicatedRegisteredClient = registeredClientBuilder
+      .id(java.util.UUID.randomUUID().toString())
+      .clientId(clientIdService.incrementClientId(client.clientId))
+      .clientIdIssuedAt(java.time.Instant.now())
+      .clientSecret(oAuthClientSecret.encode(externalClientSecret))
+      .build()
+
+    registeredClientRepository.save(duplicatedRegisteredClient)
+    return ClientCredentialsRegistrationResponse(duplicatedRegisteredClient.clientId, externalClientSecret)
```

Comment:
Create and use DuplicateRegistrationResponse data class

Reply:
Updated